### PR TITLE
Richer Contacts API

### LINKED ISSUE #18: Request: add support for additional contact information using the Contact API
See https://github.com/exponentjs/exponent/blob/master/ios/Exponent/Versioned/Modules/Api/EXContacts.m#L56

I would like to have the ability to import additional contact information that returns data similar to this library: https://github.com/rt2zz/react-native-contacts

```{
  recordID: 1,
  company: "",
  emailAddresses: [{
    label: "work",
    email: "carl-jung@example.com",
  }],
  familyName: "Jung",
  givenName: "Carl",
  jobTitle: "",
  middleName: "",
  phoneNumbers: [{
    label: "mobile",
    number: "(555) 555-5555",
  }],
  thumbnailPath: "",
}```

Thanks!

### PR DESCRIPTION
**Pull request by grabbou**
Fixes #18. Implements richer contacts API.

Here's the current preview of what's outputted: https://gist.github.com/grabbou/5272743f4b0c08cb3279f28d4996b21f

Tested with missing street/city and other properties on iOS.

### CODE REVIEW

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m [hunk 1]
```diff
@@ -70,31 +68,60 @@ - (void)_getContactsWithPermissionGrantedAsync: (NSArray *)fields addressBook:(A
     ABRecordRef person = CFArrayGetValueAtIndex(allPeople, index);
     contact[@"id"] = @(ABRecordGetRecordID(person));
     
-    NSString *firstName = (__bridge_transfer NSString *)(ABRecordCopyValue(person, kABPersonFirstNameProperty));
-    NSString *lastName = (__bridge_transfer NSString *)(ABRecordCopyValue(person, kABPersonLastNameProperty));
-    NSString *name = [self _assembleDisplayNameFromFirstName:firstName lastName:lastName];
-    if (name) {
-      contact[@"name"] = name;
```
Comment (grabbou) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Not sure if we need to check for value, AFAIK passing `null` should work via bridge.

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m [hunk 2]
```diff
@@ -53,48 +53,77 @@ - (void)dealloc
...
+      
+      if (numberOfEmails > 0) {
+        contact[@"emails"] = [NSMutableArray new];
+        
+        for (NSUInteger index = 0; index < numberOfEmails; index++) {
+          NSString *emailAddress = (__bridge_transfer NSString *)(ABMultiValueCopyValueAtIndex(emails, index));
+          
+          CFStringRef emailLabelRef = ABMultiValueCopyLabelAtIndex(emails, index);
+          NSString *emailLabel = emailLabelRef ?
+            (__bridge_transfer NSString *)(ABAddressBookCopyLocalizedLabel(emailLabelRef)) :
```
Comment (grabbou) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: I've been always indenting on tenary, like here. Xcode keeps aligning that line with `NSString` on 114, is that against our conventions?

**Comment by satya164** [comment]
Instead of a `phoneNumber` field and a `phoneNumbers` field, does it make more sense to to add a field called `default: true` to the default phone number?

**Comment by satya164** [comment]
> I am not really sure why we need to pass fields array to the native side

Last time I tried on Android, fetching contacts was very slow. Maybe because of that?

**Comment by grabbou** [comment]
@satya164 I guess the default number is first item from an array? Anyways, I removed `phoneNumber` since we have an array. Not sure if returning first item from it separately makes sense tho

**Comment by satya164** [comment]
@grabbou What if there is no default though?

**Comment by grabbou** [comment]
What you mean? AFAIK there's no explicit default, at least on iOS. That's why Exponent used to return `phoneNumber` as a first item from an array (at least on iOS). I removed that and replaced in favour of an array - which seems more reasonable. Initially I kept the `phoneNumber` as is to provide non-breaking change, but Brent said it's irrelevant and easy to upgrade, so lets' do it

**Comment by satya164** [comment]
@grabbou on Android you can set a number as default

**Comment by grabbou** [comment]
Ok, you are right, then I should either use flag or `phoneNumber` (I guess latter makes more sense). I'll check android and play around with naming. Thanks!

**Comment by terribleben** [comment]
Thanks guys!

**Comment by brentvatne** [comment]
@grabbou @satya164 - This appears to be missing image data [ABPersonCopyImageData](http://stackoverflow.com/questions/2085959/get-image-of-a-person-from-iphone-address-book) -- was that intentional? Not sure if a similar thing is possible on Android. Also, not sure if Android can return `firstName`, `middleName`, and `lastName` but that would be good for consistency.

**Comment by grabbou** [comment]
We delayed implementation of an image to next iteration (aka follow-up PR) to get this out faster (there was a deadline of a day I guess before the release).

**Comment by satya164** [comment]
@brentvatne 

> This appears to be missing image data ABPersonCopyImageData

What's structure of the data JS expects? On Android we can get URI of the person's avatar

> Also, not sure if Android can return firstName, middleName, and lastName but that would be good for consistency

Can't, android only has display name (also that's what we enter when saving contacts). Do people have to enter these 3 separate fields on iOS when adding a new contact?

**Comment by brentvatne** [comment]
@satya164

> What's structure of the data JS expects? On Android we can get URI of the person's avatar

I think the URI is fine!

> Can't, android only has display name (also that's what we enter when saving contacts). Do people have to enter these 3 separate fields on iOS when adding a new contact?

Ah ok! I'm not sure how to add the middle name on iOS but I guess it must be possible somehow. First name and last name are separate though.

**Pull request closed without merging**